pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java

```diff
@@ -831,6 +777,156 @@ private String getGroupKey(IntArray rawKey) {
     return groupKeyBuilder.toString();
   }
 
+  /**
+   * Fast int-to-int hashmap with {@link #INVALID_ID} as the default return value.
+   * <p>Different from {@link it.unimi.dsi.fastutil.ints.Int2IntOpenHashMap}, this map uses one single array to store
+   * keys and values to reduce the cache miss.
+   */
+  @VisibleForTesting
+  public static class IntGroupIdMap {
+    private static final float LOAD_FACTOR = 0.75f;
+
+    private int[] _keyValueHolder;
+    private int _capacity;
+    private int _mask;
+    private int _maxNumEntries;
+    private int _size;
+
+    public IntGroupIdMap() {
+      _capacity = 1 << 10;
+      int holderSize = _capacity << 1;
```

Comment:
So looks like there are 2048 slots in the keyValueHolder array and we are going to use the lower order 12 bits of the  incoming key for hash computation to get to the index/slot. And we will resize when 750 slots are full. Right ? How did we come up with these values of 1024, 2048 etc

Reply:
We don't want to start with a too small map, nor a too large map. Small map will result in more hash collisions and more resizes. Large map will result in bigger memory footprint. A map of capacity 1024 is quite small, but we will expand and cache it if we need a larger map. This value does not matter that much as long as it is not too large.

Reply:
Changed the initial capacity to 512 so that the array can fit into a single memory page. Added more comments also